// force-language-services/dist/visualforcepage/main/vfPageCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const vfPageModule_1 = require("./vfPageModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
const VF_KIND = 'page';
exports.vfPageCreateCommand = {
    command: 'visualforce:page:create',
    description: Messages.get('VfCreateDescription', VF_KIND),
    longDescription: Messages.get('VfCreateLongDescription', VF_KIND, VF_KIND),
    help: Messages.get('VfCreateHelp', VF_KIND, VF_KIND, VF_KIND, VF_KIND, VF_KIND),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(vfPageModule_1.VFPageModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(vfPageModule_1.VFPageModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options);
            if (context.flags.label) {
                options['vfLabel'] = context.flags.label;
            }
            const pageModule = new vfPageModule_1.VFPageModule();
            return yield pageModule.create(context.flags[vfPageModule_1.VFPageModule.getNameFlagName(VF_KIND)], context.flags[vfPageModule_1.VFPageModule.getPathFlagName()], options);
        });
    }
};
