// salesforcedx-templates/lib/commands/force/visualforce/page/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const visualforcePageGenerator_1 = require("../../../../generators/visualforcePageGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const visualforcePageFileSuffix = /.page$/;
class VisualforcePage extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.pagename);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            return this.runGenerator(visualforcePageGenerator_1.default);
        });
    }
}
VisualforcePage.examples = [
    '$ sfdx force:visualforce:page:create -n mypage -l mylabel',
    '$ sfdx force:visualforce:page:create -n mypage -l mylabel -d pages'
];
VisualforcePage.description = messages.getMessage('VisualforcePageCommandDescription');
VisualforcePage.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        default: process.cwd()
    }),
    apiversion: command_1.flags.builtin(),
    pagename: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('visualforcepagename'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultVFPage',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(visualforcePageFileSuffix, 'visualforcepage')
    }),
    label: command_1.flags.string({
        char: 'l',
        description: messages.getMessage('pagelabel'),
        required: true
    })
};
exports.default = VisualforcePage;
