// salesforcedx-templates/lib/commands/force/lightning/app/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const path = require("path");
const lightningAppGenerator_1 = require("../../../../generators/lightningAppGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const lightningAppFileSuffix = /.app$/;
class LightningApp extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.appname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            const fileparts = path.resolve(this.flags.outputdir).split(path.sep);
            // tslint:disable-next-line:no-unused-expression
            if (!fileparts.includes('aura') && !this.flags.internal) {
                throw new Error(messages.getMessage('MissingAuraDir'));
            }
            return this.runGenerator(lightningAppGenerator_1.default);
        });
    }
}
LightningApp.examples = [
    '$ sfdx force:lightning:app:create -n myapp',
    '$ sfdx force:lightning:app:create -n myapp -d aura'
];
LightningApp.description = messages.getMessage('LightningAppCommandDescription');
LightningApp.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        default: process.cwd()
    }),
    apiversion: command_1.flags.builtin(),
    appname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('appname'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultLightningApp',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningAppFileSuffix, 'lightningapp')
    }),
    internal: command_1.flags.boolean({
        char: 'i',
        description: messages.getMessage('internal'),
        hidden: true
    })
};
exports.default = LightningApp;
