// salesforcedx-templates/lib/commands/force/apex/trigger/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const apexTriggerGenerator_1 = require("../../../../generators/apexTriggerGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const apexTriggerFileSuffix = /.trigger$/;
class ApexTrigger extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.triggername);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            return this.runGenerator(apexTriggerGenerator_1.default);
        });
    }
}
ApexTrigger.examples = [
    '$ sfdx force:apex:trigger:create -n MyTrigger',
    "$ sfdx force:apex:trigger:create -n MyTrigger -s Account -e 'before insert, after insert'",
    '$ sfdx force:apex:trigger:create -n MyTrigger -d triggers'
];
ApexTrigger.description = messages.getMessage('ApexTriggerCommandDescription');
ApexTrigger.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        required: false,
        default: process.cwd()
    }),
    // Need to fix the apiversion flag with default and optional inputs
    apiversion: command_1.flags.builtin(),
    triggerevents: command_1.flags.string({
        char: 'e',
        description: messages.getMessage('triggerevents'),
        default: 'before insert',
        options: [
            'before insert',
            'before update',
            'before delete',
            'after insert',
            'after update',
            'after delete',
            'after undelete'
        ]
    }),
    triggername: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('triggername'),
        required: true
    }),
    sobject: command_1.flags.string({
        char: 's',
        description: messages.getMessage('sobject'),
        default: 'SOBJECT'
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'ApexTrigger',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(apexTriggerFileSuffix, 'apextrigger')
    })
};
exports.default = ApexTrigger;
