// force-language-services/dist/lightningtest/main/lightningTestGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class LightningTestGeneratorInfo {
    getSuffix() {
        return LightningTestGeneratorInfo.LIGHTNING_TEST_SUFFIX;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(__dirname, LightningTestGeneratorInfo.LIGHTNING_TEST_SUFFIX);
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getApiVersions() {
        return generatorUtil_1.GeneratorUtil.getValidApiVersionValues(LightningTestGeneratorInfo.LOW_API_VERSION, LightningTestGeneratorInfo.HIGH_API_VERSION);
    }
    getDefaultApiVersion() {
        return LightningTestGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getHighApiVersion() {
        return LightningTestGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getLowApiVersion() {
        return LightningTestGeneratorInfo.LOW_API_VERSION.toFixed(1);
    }
    isValidApiVersion(apiVersion) {
        return generatorUtil_1.GeneratorUtil.isValidApiVersion(apiVersion, LightningTestGeneratorInfo.LOW_API_VERSION, LightningTestGeneratorInfo.HIGH_API_VERSION);
    }
}
LightningTestGeneratorInfo.LIGHTNING_TEST_SUFFIX = '.resource';
LightningTestGeneratorInfo.LIGHTNING_TEST_NAME_OPTION = 'apiName';
LightningTestGeneratorInfo.LOW_API_VERSION = configuration_1.Configuration.getDefaultLowApiVersion();
LightningTestGeneratorInfo.HIGH_API_VERSION = configuration_1.Configuration.getDefaultHighApiVersion();
exports.LightningTestGeneratorInfo = LightningTestGeneratorInfo;
