// force-language-services/dist/lightningtest/main/lightningTestGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generator = require("yeoman-generator");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const lightningTestGeneratorInfo_1 = require("./lightningTestGeneratorInfo");
const Messages = require("../../common/main/Messages");
const lightningTestModule_1 = require("./lightningTestModule");
class LightningTestGenerator extends generator.Base {
    constructor(args, options) {
        super(args, options);
        this.genInfo = new lightningTestGeneratorInfo_1.LightningTestGeneratorInfo();
        generatorUtil_1.GeneratorUtil.setPaths(__dirname, this, options);
        this.argument(lightningTestGeneratorInfo_1.LightningTestGeneratorInfo.LIGHTNING_TEST_NAME_OPTION, {
            desc: Messages.get('LightningTestNameOption'),
            required: false,
            type: 'String',
            defaults: ''
        });
    }
    prompting() {
        let that = this;
        function x(answer) {
            that.apiName = answer['apiName'];
        }
        if (!this.apiName) {
            return this.prompt({
                type: 'input',
                name: 'apiName',
                message: Messages.get('LightningTestNamePrompt')
            }).then(answer => x(answer)); // type coercion due to the weak .d.ts for the base generator
        }
    }
    writing() {
        this.fs.copyTpl(this.templatePath(this.genInfo.getDefaultTemplate() + this.genInfo.getSuffix()), this.destinationPath(this.apiName + this.genInfo.getSuffix()), {
            apiName: this.apiName
        });
        const isInternal = this.options[lightningTestModule_1.LightningTestModule.LIGHTNING_INTERNAL_FLAG];
        if (typeof isInternal === undefined || !isInternal) {
            this.fs.copyTpl(this.templatePath('_test.resource-meta.xml'), this.destinationPath(this.apiName + '.resource-meta.xml'), {
                apiName: this.apiName,
            });
        }
    }
}
exports.LightningTestGenerator = LightningTestGenerator;
;
