// force-language-services/dist/lightningtest/main/lightningTestCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const lightningTestModule_1 = require("./lightningTestModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.lightningTestCreateCommand = {
    command: 'lightning:test:create',
    description: Messages.get('LightningTestCreateDescription'),
    longDescription: Messages.get('LightningTestCreateLongDescription'),
    help: Messages.get('LightningTestCreateHelp'),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(lightningTestModule_1.LightningTestModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(lightningTestModule_1.LightningTestModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options, lightningTestModule_1.LightningTestModule.getFlagMap());
            const lightningTest = new lightningTestModule_1.LightningTestModule();
            return yield lightningTest.create(context.flags[lightningTestModule_1.LightningTestModule.getNameFlagName()], context.flags[lightningTestModule_1.LightningTestModule.getPathFlagName()], options);
        });
    }
};
