// force-language-services/dist/apex/main/apexClassGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generator = require("yeoman-generator");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const apexGeneratorInfo_1 = require("./apexGeneratorInfo");
const Messages = require("../../common/main/Messages");
class ApexClassGenerator extends generator.Base {
    constructor(args, options) {
        super(args, options);
        this.genInfo = new apexGeneratorInfo_1.ApexGeneratorInfo();
        generatorUtil_1.GeneratorUtil.setPaths(__dirname, this, options);
        this.argument(apexGeneratorInfo_1.ApexGeneratorInfo.APEX_NAME_OPTION, {
            desc: Messages.get('ApexNameOption'),
            required: false,
            type: 'String',
            defaults: ''
        });
        this.option(generatorUtil_1.GeneratorUtil.API_VERSION_OPTION, {
            desc: Messages.get('ApiVersionOption'),
            type: 'String',
            defaults: this.genInfo.getDefaultApiVersion()
        });
        this.option(generatorUtil_1.GeneratorUtil.TEMPLATE_OPTION, {
            desc: Messages.get('TemplateOption'),
            type: 'String',
            defaults: this.genInfo.getDefaultTemplate()
        });
    }
    prompting() {
        let that = this;
        function x(answer) {
            that.apiName = answer['apiName'];
        }
        if (!this.apiName) {
            return this.prompt({
                type: 'input',
                name: 'apiName',
                message: Messages.get('ApexClassNamePrompt')
            }).then(answer => x(answer)); // type coercion due to the weak .d.ts for the base generator
        }
    }
    writing() {
        this.fs.copyTpl(this.templatePath(this.options['template'] + this.genInfo.getSuffix()), this.destinationPath(this.apiName + this.genInfo.getSuffix()), {
            apiName: this.apiName
        });
        this.fs.copyTpl(this.templatePath('_class.cls-meta.xml'), this.destinationPath(this.apiName + '.cls-meta.xml'), {
            apiName: this.apiName,
            apiVersion: this.options['apiVersion']
        });
    }
}
exports.ApexClassGenerator = ApexClassGenerator;
;
