// force-language-services/dist/apex/main/apexGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class ApexGeneratorInfo {
    getSuffix() {
        return ApexGeneratorInfo.APEX_SUFFIX;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(__dirname, ApexGeneratorInfo.APEX_SUFFIX);
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getApiVersions() {
        return generatorUtil_1.GeneratorUtil.getValidApiVersionValues(ApexGeneratorInfo.LOW_API_VERSION, ApexGeneratorInfo.HIGH_API_VERSION);
    }
    getDefaultApiVersion() {
        return ApexGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getHighApiVersion() {
        return ApexGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getLowApiVersion() {
        return ApexGeneratorInfo.LOW_API_VERSION.toFixed(1);
    }
    isValidApiVersion(apiVersion) {
        return generatorUtil_1.GeneratorUtil.isValidApiVersion(apiVersion, ApexGeneratorInfo.LOW_API_VERSION, ApexGeneratorInfo.HIGH_API_VERSION);
    }
}
ApexGeneratorInfo.APEX_SUFFIX = '.cls';
ApexGeneratorInfo.APEX_NAME_OPTION = 'apiName';
ApexGeneratorInfo.LOW_API_VERSION = configuration_1.Configuration.getDefaultLowApiVersion();
ApexGeneratorInfo.HIGH_API_VERSION = configuration_1.Configuration.getDefaultHighApiVersion();
exports.ApexGeneratorInfo = ApexGeneratorInfo;
