// force-language-services/dist/apex/main/apexCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const apexModule_1 = require("./apexModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.apexClassCreateCommand = {
    command: 'apex:class:create',
    description: Messages.get('ApexCreateDescription'),
    longDescription: Messages.get('ApexCreateLongDescription'),
    help: Messages.get('ApexCreateHelp'),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(apexModule_1.ApexModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(apexModule_1.ApexModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options);
            const apex = new apexModule_1.ApexModule();
            return yield apex.create(context.flags[apexModule_1.ApexModule.getNameFlagName()], context.flags[apexModule_1.ApexModule.getPathFlagName()], options);
        });
    }
};
