// force-language-services/dist/project/main/projectCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const projectModule_1 = require("./projectModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.projectCreateCommand = {
    command: 'project:create',
    description: Messages.get('ProjectCreateCommandDescription'),
    longDescription: Messages.get('ProjectCreateCommandLongDescription'),
    help: Messages.get('ProjectCreateCommandHelp'),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(projectModule_1.ProjectModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(projectModule_1.ProjectModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options, projectModule_1.ProjectModule.getFlagMap());
            const workspace = new projectModule_1.ProjectModule();
            return yield workspace.create(context.flags[projectModule_1.ProjectModule.getNameFlagName()], context.flags[projectModule_1.ProjectModule.getPathFlagName()], options);
        });
    }
};
