// force-language-services/dist/project/main/projectGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class ProjectGeneratorInfo {
    getSuffix() {
        return ProjectGeneratorInfo.PROJECT_SUFFIX;
    }
    getTemplates() {
        return [
            ProjectGeneratorInfo.STANDARD_TEMPLATE,
            ProjectGeneratorInfo.EMPTY_TEMPLATE
        ];
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getDefaultLoginURL() {
        return ProjectGeneratorInfo.DEFAULT_LOGIN_URL;
    }
    getDefaultArtifactName() {
        return ProjectGeneratorInfo.DEFAULT_PROJECT_NAME;
    }
    getDefaultSourceApiVersion() {
        return configuration_1.Configuration.getCurrentApiVersionString();
    }
    // TODO dummy or split GeneratorInfo interface
    getApiVersions() {
        let xxx = [];
        return xxx;
    }
    getDefaultApiVersion() {
        return '';
    }
    isValidApiVersion(apiVersion) {
        return false;
    }
}
ProjectGeneratorInfo.PROJECT_NAME_OPTION = 'projectName';
ProjectGeneratorInfo.LOGIN_URL_OPTION = 'loginURL';
ProjectGeneratorInfo.NAMESPACE_OPTION = 'projectNamespace';
ProjectGeneratorInfo.DEFAULT_PROJECT_OPTION = 'defaultArtifact';
ProjectGeneratorInfo.SCRATCH_ORG_DEF_OPTION = 'scratchOrgDef';
ProjectGeneratorInfo.MANIFEST_OPTION = 'manifest';
ProjectGeneratorInfo.SOURCE_API_VERSION_OPTION = 'sourceApiVersion';
ProjectGeneratorInfo.PROJECT_SUFFIX = '';
ProjectGeneratorInfo.PROJECT_CONFIG_NAME = 'sfdx-project.json';
ProjectGeneratorInfo.README = 'README.md';
ProjectGeneratorInfo.README_EMPTY = 'README.empty.md';
ProjectGeneratorInfo.README_STANDARD = 'README.standard.md';
ProjectGeneratorInfo.FORCE_IGNORE = '.forceignore';
ProjectGeneratorInfo.GIT_IGNORE = '.gitignore';
ProjectGeneratorInfo.GIT_IGNORE_TEMPLATE = 'gitignore';
ProjectGeneratorInfo.EMPTY_TEMPLATE = 'empty';
ProjectGeneratorInfo.STANDARD_TEMPLATE = 'standard';
ProjectGeneratorInfo.LEGACY_TEMPLATE = 'Defaultsfdx-project.json';
ProjectGeneratorInfo.DEFAULT_PROJECT_NAME = 'force-app';
ProjectGeneratorInfo.DEFAULT_LOGIN_URL = 'https://login.salesforce.com';
exports.ProjectGeneratorInfo = ProjectGeneratorInfo;
