// force-language-services/dist/visualforcecommon/main/vfGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class VFGeneratorInfo {
    constructor(suffix, genDir) {
        this.vfSuffix = suffix;
        this.genDir = genDir;
    }
    getSuffix() {
        return this.vfSuffix;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(this.genDir, this.getSuffix());
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getApiVersions() {
        return generatorUtil_1.GeneratorUtil.getValidApiVersionValues(VFGeneratorInfo.LOW_API_VERSION, VFGeneratorInfo.HIGH_API_VERSION);
    }
    getDefaultApiVersion() {
        return VFGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getHighApiVersion() {
        return VFGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getLowApiVersion() {
        return VFGeneratorInfo.LOW_API_VERSION.toFixed(1);
    }
    isValidApiVersion(apiVersion) {
        return generatorUtil_1.GeneratorUtil.isValidApiVersion(apiVersion, VFGeneratorInfo.LOW_API_VERSION, VFGeneratorInfo.HIGH_API_VERSION);
    }
}
VFGeneratorInfo.LOW_API_VERSION = configuration_1.Configuration.getDefaultLowApiVersion();
VFGeneratorInfo.HIGH_API_VERSION = configuration_1.Configuration.getDefaultHighApiVersion();
exports.VFGeneratorInfo = VFGeneratorInfo;
