// salesforcedx-templates/lib/commands/force/apex/class/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const apexClassGenerator_1 = require("../../../../generators/apexClassGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const apexClassFileSuffix = /.cls$/;
class ApexClass extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.classname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            return this.runGenerator(apexClassGenerator_1.default);
        });
    }
}
ApexClass.examples = [
    '$ sfdx force:apex:class:create -n MyClass',
    '$ sfdx force:apex:class:create -n MyClass -d classes'
];
ApexClass.description = messages.getMessage('ApexClassCommandDescription');
ApexClass.flagsConfig = {
    classname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('NameFlagDescription'),
        required: true
    }),
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        required: false,
        default: process.cwd()
    }),
    // Need to fix the apiversion flag with default and optional inputs
    apiversion: command_1.flags.builtin(),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultApexClass',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(apexClassFileSuffix, 'apexclass')
    })
};
exports.default = ApexClass;
