// salesforcedx-templates/lib/commands/force/visualforce/component/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const visualforceComponentGenerator_1 = require("../../../../generators/visualforceComponentGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const visualforceComponentFileSuffix = /.component$/;
class VisualforceComponent extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.componentname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            return this.runGenerator(visualforceComponentGenerator_1.default);
        });
    }
}
VisualforceComponent.examples = [
    '$ sfdx force:visualforce:component:create -n mycomponent -l mylabel',
    '$ sfdx force:visualforce:component:create -n mycomponent -l mylabel -d components'
];
VisualforceComponent.description = messages.getMessage('VisualforceComponentCommandDescription');
VisualforceComponent.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        default: process.cwd()
    }),
    apiversion: command_1.flags.builtin(),
    componentname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('visualforcecomponentname'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultVFComponent',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(visualforceComponentFileSuffix, 'visualforcecomponent')
    }),
    label: command_1.flags.string({
        char: 'l',
        description: messages.getMessage('componentlabel'),
        required: true
    })
};
exports.default = VisualforceComponent;
