// salesforcedx-templates/lib/commands/force/lightning/component/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const path = require("path");
const lightningComponentGenerator_1 = require("../../../../generators/lightningComponentGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const lightningComponentFileSuffix = /.cmp$/;
class LightningComponent extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.componentname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            // lwc requires first letter lowercase
            if (this.flags.type === 'lwc') {
                this.flags.componentname = `${this.flags.componentname
                    .substring(0, 1)
                    .toLowerCase()}${this.flags.componentname.substring(1)}`;
            }
            const fileparts = path.resolve(this.flags.outputdir).split(path.sep);
            // tslint:disable-next-line:no-unused-expression
            if (!this.flags.internal) {
                if (this.flags.type === 'lwc' && !fileparts.includes('lwc')) {
                    throw new Error(messages.getMessage('MissingLWCDir'));
                }
                else if (!fileparts.includes('aura') && this.flags.type === 'aura') {
                    throw new Error(messages.getMessage('MissingAuraDir'));
                }
            }
            return this.runGenerator(lightningComponentGenerator_1.default);
        });
    }
}
LightningComponent.examples = [
    '$ sfdx force:lightning:component:create -n mycomponent',
    '$ sfdx force:lightning:component:create -n mycomponent --type lwc',
    '$ sfdx force:lightning:component:create -n mycomponent -d aura',
    '$ sfdx force:lightning:component:create -n mycomponent --type lwc -d lwc'
];
LightningComponent.description = messages.getMessage('LightningComponentCommandDescription');
LightningComponent.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        required: false,
        default: process.cwd()
    }),
    apiversion: command_1.flags.builtin(),
    componentname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('componentname'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultLightningCmp',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningComponentFileSuffix, 'lightningcomponent')
    }),
    type: command_1.flags.string({
        description: messages.getMessage('ComponentType'),
        options: ['aura', 'lwc'],
        default: 'aura'
    }),
    internal: command_1.flags.boolean({
        char: 'i',
        description: messages.getMessage('internal'),
        hidden: true
    })
};
exports.default = LightningComponent;
