// force-language-services/dist/manifest/main/manifestModule.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const manifestGenerator_1 = require("./manifestGenerator");
const createBase_1 = require("../../common/main/createBase");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const commandUtil_1 = require("../../common/main/commandUtil");
const manifestGeneratorInfo_1 = require("./manifestGeneratorInfo");
const Messages = require("../../common/main/Messages");
const projectModule_1 = require("../../project/main/projectModule");
const projectGeneratorInfo_1 = require("../../project/main/projectGeneratorInfo");
class ManifestModule {
    static getFlagMap() {
        let flagMap = new Map();
        flagMap.set(projectModule_1.ProjectModule.SOURCE_API_VERSION_FLAG, projectGeneratorInfo_1.ProjectGeneratorInfo.SOURCE_API_VERSION_OPTION);
        return flagMap;
    }
    create(manifestName, targetPath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const generatorName = 'force:manifest';
            let addGenerators = function (env) {
                env.registerStub(manifestGenerator_1.ManifestGenerator, generatorName);
            };
            let checkInputs = function () {
                let templateName = options['template'];
                if (templateName) {
                    if (!ManifestModule.genInfo.isValidTemplate(templateName)) {
                        return Messages.get('InvalidTemplate');
                    }
                }
                return '';
            };
            let createOptions = options;
            createOptions[manifestGeneratorInfo_1.ManifestGeneratorInfo.MANIFEST_NAME_OPTION] = manifestName;
            let rawOutput = yield new createBase_1.CreateBase().create(manifestName, targetPath, createOptions, generatorName, checkInputs, addGenerators);
            if (options['json']) {
                return generatorUtil_1.GeneratorUtil.jsonOutput(rawOutput);
            }
            else {
                return rawOutput;
            }
        });
    }
    static getNameFlagName() {
        return 'manifestname' + commandUtil_1.CommandUtil.NAME_FLAG;
    }
    static getPathFlagName() {
        return commandUtil_1.CommandUtil.PATH_FLAG;
    }
    static reflectCreate() {
        let reflect = { flags: {} };
        generatorUtil_1.GeneratorUtil.addNameFlag(reflect, ManifestModule.getNameFlagName(), Messages.get('ManifestNameFlagDescription'), Messages.get('ManifestNameFlagLongDescription'));
        generatorUtil_1.GeneratorUtil.addCommonFlags2(reflect, ManifestModule.genInfo.getTemplates(), ManifestModule.genInfo.getDefaultTemplate());
        generatorUtil_1.GeneratorUtil.addFlag(reflect, {
            name: projectModule_1.ProjectModule.SOURCE_API_VERSION_FLAG,
            char: 'x',
            hasValue: true,
            required: false,
            hidden: true,
            description: Messages.get('ProjectVersionFlagDescription'),
            longDescription: Messages.get('ProjectVersionFlagLongDescription')
        }, [], projectModule_1.ProjectModule.genInfo.getDefaultSourceApiVersion());
        return reflect;
    }
}
ManifestModule.genInfo = new manifestGeneratorInfo_1.ManifestGeneratorInfo();
exports.ManifestModule = ManifestModule;
