// force-language-services/dist/manifest/main/manifestGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class ManifestGeneratorInfo {
    getSuffix() {
        return ManifestGeneratorInfo.MANIFEST_SUFFIX;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(__dirname, ManifestGeneratorInfo.MANIFEST_SUFFIX);
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    getDefaultSourceApiVersion() {
        return configuration_1.Configuration.getCurrentApiVersionString();
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
}
ManifestGeneratorInfo.MANIFEST_CONFIG_FOLDER = 'manifest';
ManifestGeneratorInfo.MANIFEST_SUFFIX = '.xml';
ManifestGeneratorInfo.DEFAULT_MANIFEST_NAME = 'package.xml';
ManifestGeneratorInfo.MANIFEST_NAME_OPTION = 'manifestNameOption';
ManifestGeneratorInfo.SOURCE_API_VERSION_OPTION = 'sourceApiVersion';
exports.ManifestGeneratorInfo = ManifestGeneratorInfo;
