// salesforcedx-templates/lib/commands/force/lightning/test/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const lightningTestGenerator_1 = require("../../../../generators/lightningTestGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const lightningTestFileSuffix = /.resource$/;
class LightningTest extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.testname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            return this.runGenerator(lightningTestGenerator_1.default);
        });
    }
}
LightningTest.examples = [
    '$ sfdx force:lightning:test:create -n MyLightningTest',
    '$ sfdx force:lightning:test:create -n MyLightningTest -d lightningTests'
];
LightningTest.description = messages.getMessage('LightningTestCommandDescription');
LightningTest.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        required: false,
        default: process.cwd()
    }),
    testname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('testname'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultLightningTest',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningTestFileSuffix, 'lightningtest')
    }),
    internal: command_1.flags.boolean({
        char: 'i',
        description: messages.getMessage('internal'),
        hidden: true
    })
};
exports.default = LightningTest;
