// force-language-services/dist/visualforcecomponent/main/vfComponentCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const vfComponentModule_1 = require("./vfComponentModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
const VF_KIND = 'component';
exports.vfComponentCreateCommand = {
    command: 'visualforce:component:create',
    description: Messages.get('VfCreateDescription', VF_KIND),
    longDescription: Messages.get('VfCreateLongDescription', VF_KIND, VF_KIND),
    help: Messages.get('VfCreateHelp', VF_KIND, VF_KIND, VF_KIND, VF_KIND, VF_KIND),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(vfComponentModule_1.VFComponentModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(vfComponentModule_1.VFComponentModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options);
            if (context.flags.label) {
                options['vfLabel'] = context.flags.label;
            }
            const componentModule = new vfComponentModule_1.VFComponentModule();
            return yield componentModule.create(context.flags[vfComponentModule_1.VFComponentModule.getNameFlagName(VF_KIND)], context.flags[vfComponentModule_1.VFComponentModule.getPathFlagName()], options);
        });
    }
};
