// force-language-services/dist/scratchdef/main/scratchDefCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const scratchDefModule_1 = require("./scratchDefModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.scratchDefCreateCommand = {
    command: 'scratchdef:create',
    description: Messages.get('ScratchDefCreateCommandDescription'),
    longDescription: Messages.get('ScratchDefCreateCommandLongDescription'),
    help: Messages.get('ScratchDefCreateCommandHelp'),
    hidden: true,
    args: [],
    flags: reflection_1.getCommandFlags(scratchDefModule_1.ScratchDefModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(scratchDefModule_1.ScratchDefModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options, scratchDefModule_1.ScratchDefModule.getFlagMap());
            const scratchDef = new scratchDefModule_1.ScratchDefModule();
            return yield scratchDef.create(context.flags[scratchDefModule_1.ScratchDefModule.getNameFlagName()], context.flags[scratchDefModule_1.ScratchDefModule.getPathFlagName()], options);
        });
    }
};
