// force-language-services/dist/scratchdef/main/scratchDefGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const process = require("process");
class ScratchDefGeneratorInfo {
    static getUserName() {
        return process.env.USER || 'Demo';
    }
    getSuffix() {
        return ScratchDefGeneratorInfo.SCRATCHDEF_SUFFIX;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(__dirname, ScratchDefGeneratorInfo.SCRATCHDEF_SUFFIX);
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getDefaultCompanyName() {
        return ScratchDefGeneratorInfo.getUserName() + ' Company';
    }
    getDefaultLastName() {
        return ScratchDefGeneratorInfo.getUserName();
    }
}
ScratchDefGeneratorInfo.SCRATCHDEF_SUFFIX = '.json';
ScratchDefGeneratorInfo.SCRATCHDEF_CONFIG_FOLDER = 'config';
ScratchDefGeneratorInfo.DEFAULT_SCRATCHDEF_NAME = 'project-scratch-def.json';
ScratchDefGeneratorInfo.SCRATCH_DEF_NAME_OPTION = 'scratchDefName';
ScratchDefGeneratorInfo.COMPANY_OPTION = 'company';
ScratchDefGeneratorInfo.LAST_NAME_OPTION = 'lastName';
ScratchDefGeneratorInfo.EMAIL_OPTION = 'email';
exports.ScratchDefGeneratorInfo = ScratchDefGeneratorInfo;
