// force-language-services/dist/prettier/main/prettierCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const prettierModule_1 = require("./prettierModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.prettierCreateCommand = {
    command: 'prettier:create',
    description: Messages.get('PrettierCreateCommandDescription'),
    longDescription: Messages.get('PrettierCreateCommandLongDescription'),
    help: Messages.get('PrettierCreateCommandHelp'),
    hidden: true,
    args: [],
    flags: reflection_1.getCommandFlags(prettierModule_1.PrettierModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(prettierModule_1.PrettierModule.reflectCreate());
            }
            commandUtil_1.translateContextFlagsToOptions(context, options, prettierModule_1.PrettierModule.getFlagMap());
            const vsCodeModule = new prettierModule_1.PrettierModule();
            return yield vsCodeModule.create(context.flags[prettierModule_1.PrettierModule.getPathFlagName()], options);
        });
    }
};
