// force-language-services/dist/eslint/main/eslintCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const eslintModule_1 = require("./eslintModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.eslintCreateCommand = {
    command: 'eslint:create',
    description: Messages.get('ESLintCreateCommandDescription'),
    longDescription: Messages.get('ESLintCreateCommandLongDescription'),
    help: Messages.get('ESLintCreateCommandHelp'),
    hidden: true,
    args: [],
    flags: reflection_1.getCommandFlags(eslintModule_1.ESLintModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(eslintModule_1.ESLintModule.reflectCreate());
            }
            commandUtil_1.translateContextFlagsToOptions(context, options, eslintModule_1.ESLintModule.getFlagMap());
            const eslintModule = new eslintModule_1.ESLintModule();
            return yield eslintModule.create(context.flags[eslintModule_1.ESLintModule.getPathFlagName()], options);
        });
    }
};
