// force-language-services/dist/eslint/main/eslintGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generator = require("yeoman-generator");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const eslintGeneratorInfo_1 = require("./eslintGeneratorInfo");
const Messages = require("../../common/main/Messages");
class ESLintGenerator extends generator.Base {
    constructor(args, options) {
        super(args, options);
        this.genInfo = new eslintGeneratorInfo_1.ESLintGeneratorInfo();
        generatorUtil_1.GeneratorUtil.setPaths(__dirname, this, options);
        this.option(generatorUtil_1.GeneratorUtil.TEMPLATE_OPTION, {
            desc: Messages.get('TemplateOption'),
            type: 'String',
            defaults: this.genInfo.getDefaultTemplate()
        });
    }
    writing() {
        this.fs.copyTpl(this.templatePath(this.options[generatorUtil_1.GeneratorUtil.TEMPLATE_OPTION] + this.genInfo.getSuffix()), this.destinationPath(eslintGeneratorInfo_1.ESLintGeneratorInfo.ESLINT_RC), {});
    }
}
ESLintGenerator.GENERATOR_NAME = 'force:eslint';
exports.ESLintGenerator = ESLintGenerator;
