// force-language-services/dist/lightning/main/lightningBaseGeneratorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generatorUtil_1 = require("../../common/main/generatorUtil");
const configuration_1 = require("../../common/main/configuration");
class LightningBaseGeneratorInfo {
    constructor(suffix, genDir) {
        this.lSuffix = '.' + suffix;
        this.genDir = genDir;
    }
    getSuffix() {
        return this.lSuffix;
    }
    getTemplates() {
        return generatorUtil_1.GeneratorUtil.getTemplates(this.genDir, this.lSuffix);
    }
    getDefaultTemplate() {
        return this.getTemplates()[0];
    }
    isValidTemplate(template) {
        return generatorUtil_1.GeneratorUtil.isValidTemplate(this.getTemplates(), template);
    }
    getApiVersions() {
        return generatorUtil_1.GeneratorUtil.getValidApiVersionValues(LightningBaseGeneratorInfo.LOW_API_VERSION, LightningBaseGeneratorInfo.HIGH_API_VERSION);
    }
    getDefaultApiVersion() {
        return LightningBaseGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getHighApiVersion() {
        return LightningBaseGeneratorInfo.HIGH_API_VERSION.toFixed(1);
    }
    getLowApiVersion() {
        return LightningBaseGeneratorInfo.LOW_API_VERSION.toFixed(1);
    }
    isValidApiVersion(apiVersion) {
        return generatorUtil_1.GeneratorUtil.isValidApiVersion(apiVersion, LightningBaseGeneratorInfo.LOW_API_VERSION, LightningBaseGeneratorInfo.HIGH_API_VERSION);
    }
}
LightningBaseGeneratorInfo.LOW_API_VERSION = configuration_1.Configuration.getDefaultLowApiVersion();
LightningBaseGeneratorInfo.HIGH_API_VERSION = configuration_1.Configuration.getDefaultHighApiVersion();
exports.LightningBaseGeneratorInfo = LightningBaseGeneratorInfo;
