// force-language-services/dist/artifact/main/artifactCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const artifactModule_1 = require("./artifactModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.artifactCreateCommand = {
    command: 'artifact:create',
    description: Messages.get('ArtifactCreateCommandDescription'),
    longDescription: Messages.get('ArtifactCreateCommandLongDescription'),
    help: Messages.get('ArtifactCreateCommandHelp'),
    hidden: true,
    args: [],
    flags: reflection_1.getCommandFlags(artifactModule_1.ArtifactModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(artifactModule_1.ArtifactModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options, artifactModule_1.ArtifactModule.getFlagMap());
            const artifact = new artifactModule_1.ArtifactModule();
            return yield artifact.create(context.flags[artifactModule_1.ArtifactModule.getNameFlagName()], context.flags[artifactModule_1.ArtifactModule.getPathFlagName()], options);
        });
    }
};
