// force-language-services/dist/vscode/main/vscodeCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscodeModule_1 = require("./vscodeModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.vscodeCreateCommand = {
    command: 'vscode:create',
    description: Messages.get('VSCodeCreateCommandDescription'),
    longDescription: Messages.get('VSCodeCreateCommandLongDescription'),
    help: Messages.get('VSCodeCreateCommandHelp'),
    hidden: true,
    args: [],
    flags: reflection_1.getCommandFlags(vscodeModule_1.VSCodeModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(vscodeModule_1.VSCodeModule.reflectCreate());
            }
            commandUtil_1.translateContextFlagsToOptions(context, options, vscodeModule_1.VSCodeModule.getFlagMap());
            const vsCodeModule = new vscodeModule_1.VSCodeModule();
            return yield vsCodeModule.create(context.flags[vscodeModule_1.VSCodeModule.getPathFlagName()], options);
        });
    }
};
