// force-language-services/dist/apextrigger/main/apexTriggerGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generator = require("yeoman-generator");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const apexTriggerGeneratorInfo_1 = require("./apexTriggerGeneratorInfo");
const Messages = require("../../common/main/Messages");
class ApexTriggerGenerator extends generator.Base {
    constructor(args, options) {
        super(args, options);
        this.genInfo = new apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo();
        generatorUtil_1.GeneratorUtil.setPaths(__dirname, this, options);
        this.argument(apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.APEX_TRIGGER_NAME_OPTION, {
            desc: Messages.get('ApexTriggerNameOption'),
            required: false,
            type: 'String',
            defaults: ''
        });
        this.option(generatorUtil_1.GeneratorUtil.API_VERSION_OPTION, {
            desc: Messages.get('ApiVersionOption'),
            type: 'String',
            defaults: this.genInfo.getDefaultApiVersion()
        });
        this.option(generatorUtil_1.GeneratorUtil.TEMPLATE_OPTION, {
            desc: Messages.get('TemplateOption'),
            type: 'String',
            defaults: this.genInfo.getDefaultTemplate()
        });
        this.option(apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.APEX_TRIGGER_SOBJECT_OPTION, {
            desc: Messages.get('ApexTriggerSobjectOption'),
            type: 'String',
            defaults: apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.DEFAULT_SOBJECT
        });
        this.option(apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.APEX_TRIGGER_TRIGGER_EVENTS_OPTION, {
            desc: Messages.get('ApexTriggerTriggerEventsOption'),
            type: 'String',
            defaults: apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.TRIGGER_EVENTS[0]
        });
    }
    prompting() {
        let that = this;
        function x(answer) {
            that.apiName = answer['apiName'];
        }
        if (!this.apiName) {
            return this.prompt({
                type: 'input',
                name: 'apiName',
                message: Messages.get('ApexTriggerNamePrompt')
            }).then(answer => x(answer)); // type coercion due to the weak .d.ts for the base generator
        }
    }
    writing() {
        this.fs.copyTpl(this.templatePath(this.options['template'] + this.genInfo.getSuffix()), this.destinationPath(this.apiName + this.genInfo.getSuffix()), {
            apiName: this.apiName,
            sobject: this.options['sobject'],
            triggerEvents: this.options['triggerEvents']
        });
        this.fs.copyTpl(this.templatePath('_trigger.trigger-meta.xml'), this.destinationPath(this.apiName + '.trigger-meta.xml'), {
            apiName: this.apiName,
            apiVersion: this.options['apiVersion']
        });
    }
}
exports.ApexTriggerGenerator = ApexTriggerGenerator;
;
