// force-language-services/dist/apextrigger/main/apexTriggerCreateCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const apexTriggerModule_1 = require("./apexTriggerModule");
const commandUtil_1 = require("../../common/main/commandUtil");
const reflection_1 = require("../../common/main/reflection");
const Messages = require("../../common/main/Messages");
exports.apexTriggerCreateCommand = {
    command: 'apex:trigger:create',
    description: Messages.get('ApexTriggerCreateDescription'),
    longDescription: Messages.get('ApexTriggerCreateLongDescription'),
    help: Messages.get('ApexTriggerCreateHelp'),
    hidden: false,
    args: [],
    flags: reflection_1.getCommandFlags(apexTriggerModule_1.ApexTriggerModule.reflectCreate()),
    supportedGlobals: [],
    requiresProject: false,
    run: function (context) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = {};
            if (context.flags.reflect) {
                return JSON.stringify(apexTriggerModule_1.ApexTriggerModule.reflectCreate());
            }
            ;
            commandUtil_1.translateContextFlagsToOptions(context, options);
            if (context.flags.sobject) {
                options['sobject'] = context.flags.sobject;
            }
            if (context.flags.triggerevents) {
                options['triggerEvents'] = context.flags.triggerevents;
            }
            const apex = new apexTriggerModule_1.ApexTriggerModule();
            return yield apex.create(context.flags[apexTriggerModule_1.ApexTriggerModule.getNameFlagName()], context.flags[apexTriggerModule_1.ApexTriggerModule.getPathFlagName()], options);
        });
    }
};
