// force-language-services/dist/apextrigger/main/apexTriggerModule.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const apexTriggerGenerator_1 = require("./apexTriggerGenerator");
const createBase_1 = require("../../common/main/createBase");
const generatorUtil_1 = require("../../common/main/generatorUtil");
const commandUtil_1 = require("../../common/main/commandUtil");
const apexTriggerGeneratorInfo_1 = require("./apexTriggerGeneratorInfo");
const Messages = require("../../common/main/Messages");
class ApexTriggerModule {
    /**
     * @param {apiName} - name of the new apex class
     * @param {targetPath} - where to generate the apex
     * @param (options) - available options (to be documented)
     *     should allow config - which is a json file with all the options in a decision tree form)
     */
    // NOTE: the yeoman generator changes the actual cwd when the destinationRoot is set
    //       and too many things in there assume that
    //       so if there is a target path, reset the cwd after done running the generator
    create(apiName, targetPath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const generatorName = 'force:apex';
            let addGenerators = function (env) {
                env.registerStub(apexTriggerGenerator_1.ApexTriggerGenerator, generatorName);
            };
            let checkInputs = function () {
                let templateName = options['template'];
                let apiVersion = options['apiVersion'];
                let triggerEvents = options['triggerEvents'];
                const checkNameResult = generatorUtil_1.GeneratorUtil.checkName(apiName);
                if (checkNameResult.length > 0) {
                    return checkNameResult;
                }
                if (templateName) {
                    if (!ApexTriggerModule.genInfo.isValidTemplate(templateName)) {
                        return Messages.get('InvalidTemplate');
                    }
                }
                if (apiVersion) {
                    if (!ApexTriggerModule.genInfo.isValidApiVersion(apiVersion)) {
                        return Messages.get('InvalidApiVersion');
                    }
                }
                if (triggerEvents) {
                    triggerEvents = triggerEvents.replace(', ', ',');
                    let eventsSplit = triggerEvents.split(',');
                    for (let i = 0; i < eventsSplit.length; i++) {
                        if (apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.TRIGGER_EVENTS.indexOf(eventsSplit[i]) === -1) {
                            return Messages.get('InvalidTriggerEvent', eventsSplit[i]);
                        }
                    }
                }
                return '';
            };
            let createOptions = options;
            createOptions['apiName'] = apiName;
            let rawOutput = yield new createBase_1.CreateBase().create(apiName, targetPath, createOptions, generatorName, checkInputs, addGenerators);
            if (options['json']) {
                return generatorUtil_1.GeneratorUtil.jsonOutput(rawOutput);
            }
            else {
                return rawOutput;
            }
        });
    }
    static getNameFlagName() {
        return 'trigger' + commandUtil_1.CommandUtil.NAME_FLAG;
    }
    static getPathFlagName() {
        return commandUtil_1.CommandUtil.PATH_FLAG;
    }
    static reflectCreate() {
        let reflect = { flags: {} };
        generatorUtil_1.GeneratorUtil.addNameFlag(reflect, ApexTriggerModule.getNameFlagName(), Messages.get('ApexTriggerNameFlagDescription'), Messages.get('ApexTriggerNameFlagLongDescription'));
        generatorUtil_1.GeneratorUtil.addCommonFlags(reflect, ApexTriggerModule.genInfo);
        generatorUtil_1.GeneratorUtil.addApiVersionFlag(reflect, ApexTriggerModule.genInfo);
        generatorUtil_1.GeneratorUtil.addFlag(reflect, {
            name: 'sobject', char: 's', hasValue: true, required: false,
            description: Messages.get('ApexTriggerSobjectFlagDescription'),
            longDescription: Messages.get('ApexTriggerSobjectFlagLongDescription'),
            type: 'string'
        }, [], apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.DEFAULT_SOBJECT);
        generatorUtil_1.GeneratorUtil.addFlag(reflect, {
            name: 'triggerevents', char: 'e', hasValue: true, required: false,
            description: Messages.get('ApexTriggerEventsFlagDescription'),
            longDescription: Messages.get('ApexTriggerEventsFlagLongDescription'),
            type: 'string',
            array: true
        }, apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.TRIGGER_EVENTS, apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo.TRIGGER_EVENTS[0]);
        return reflect;
    }
}
ApexTriggerModule.genInfo = new apexTriggerGeneratorInfo_1.ApexTriggerGeneratorInfo();
exports.ApexTriggerModule = ApexTriggerModule;
;
