// salesforcedx-templates/lib/commands/force/lightning/event/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const path = require("path");
const lightningEventGenerator_1 = require("../../../../generators/lightningEventGenerator");
const utils_1 = require("../../../../utils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforcedx-templates', 'messages');
const lightningEventFileSuffix = /.evt$/;
class LightningEvent extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            utils_1.CreateUtil.checkInputs(this.flags.eventname);
            utils_1.CreateUtil.checkInputs(this.flags.template);
            const fileparts = path.resolve(this.flags.outputdir).split(path.sep);
            // tslint:disable-next-line:no-unused-expression
            if (!this.flags.internal && !fileparts.includes('aura')) {
                throw new Error(messages.getMessage('MissingAuraDir'));
            }
            return this.runGenerator(lightningEventGenerator_1.default);
        });
    }
}
LightningEvent.examples = [
    '$ sfdx force:lightning:app:create -n myevent',
    '$ sfdx force:lightning:event:create -n myevent -d aura'
];
LightningEvent.description = messages.getMessage('LightningEventCommandDescription');
LightningEvent.flagsConfig = {
    outputdir: command_1.flags.string({
        char: 'd',
        description: messages.getMessage('outputdir'),
        required: false,
        default: process.cwd()
    }),
    apiversion: command_1.flags.builtin(),
    eventname: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('eventname'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: messages.getMessage('template'),
        default: 'DefaultLightningEvt',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningEventFileSuffix, 'lightningevent')
    }),
    internal: command_1.flags.boolean({
        char: 'i',
        description: messages.getMessage('internal'),
        hidden: true
    })
};
exports.default = LightningEvent;
