// force-language-services/dist/common/main/Messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli = require("heroku-cli-util");
const util = require("util");
// could expand with get and set locale functions
let locale = 'en_US';
exports.get = function (label, ...args) {
    if (!messages[locale]) {
        cli.exit(1, 'Local \'' + locale + '\' doesn\'t exist');
    }
    if (!messages[locale][label]) {
        cli.exit(1, 'Message \'' + label + '\' doesn\'t exist');
    }
    if (args) {
        let expectedNumArgs = messages[locale][label].split('%s').length - 1;
        if (args.length !== expectedNumArgs) {
            cli.exit(1, 'Wrong number of args for message: ' + label + '\nExpect ' + expectedNumArgs + ' got ' + args.length);
        }
        args.unshift(messages[locale][label]);
        return util.format.apply(util, args);
    }
    return messages[locale][label];
};
const CLI_NAME = 'sfdx';
const FORCE_NAMESPACE = 'force:';
const CMD_HEAD = CLI_NAME + ' ' + FORCE_NAMESPACE;
const NAME_FLAG_BASE = 'name of the generated ';
const SF_NAME_RESTRICTIONS = 'The name can be up to 40 characters and must start with a letter.';
const HELP_DEFAULTS = 'If not supplied, the apiversion, template, and outputdir use default values.\n';
const HELP_OUTPUTDIR_RELATIVE = 'The outputdir can be an absolute path or relative to the current working directory.\n';
const HELP_OUTPUTDIR_RELATIVE_LIGHTNING = 'If you don’t specify an outputdir, we create a subfolder ' +
    'in your current working directory with the name of your bundle. For example, if the current working ' +
    'directory is force-app and your Lightning bundle is called myBundle, we create force-app/myBundle/ ' +
    'to store the files in the bundle.\n';
const HELP_EXAMPLES = '\nExamples:\n';
const LONG_DESCRIPTION_OUTPUTDIR = ' in the specified directory or the current working directory. ';
let messages = {
    en_US: {
        // common flags/options used by most commands
        ApiVersionFlag: 'API version number',
        ApiVersionFlagLongDescription: 'The API version of the created source.',
        ConfigFlag: 'JSON package of all options',
        JsonFlag: 'JSON output',
        JsonFlagLongDescription: 'Formats output as JSON.',
        ReflectionFlag: 'switch to return flag detailed information',
        ReflectionFlagLongDescription: 'Return reflection description of the command, options, and possible values. ' +
            'Allows IDE to tailor to the capabilities of the command',
        TargetPathFlag: 'folder for saving the created files',
        TargetPathLongDescription: 'The directory to store the newly created files. The location can be an absolute path or ' +
            'relative to the current working directory. The default is the current directory.',
        TemplateFlag: 'template to use for file creation',
        TemplateFlagLongDescription: 'The template to use to create the file. ' +
            'Supplied parameter values or default values are filled into a copy of the template.',
        ApiVersionOption: 'The version of the API that we want to use',
        TemplateOption: 'The template to use',
        // common errors
        InvalidTemplate: 'Specify a valid template.',
        InvalidApiVersion: 'Specify a valid API version.',
        MissingLabel: 'Provide a label.',
        MissingName: 'Provide a name.',
        MissingFlags: 'Provide all required parameters.',
        DoubleUnderscoreError: 'Name can\'t contain 2 consecutive underscores.',
        EndWithUnderscoreError: 'Name can\'t end with an underscore.',
        NameMustStartWithLetterError: 'Name must start with a letter.',
        AlphaNumericNameError: 'Name must contain only alphanumeric characters.',
        // per entity messages
        ApexCreateDescription: 'create an Apex class',
        ApexCreateLongDescription: 'Creates an Apex class' + LONG_DESCRIPTION_OUTPUTDIR +
            'If you don’t explicitly set the API version, it defaults to the current API version. ' +
            'The .cls file and associated metadata file are created.',
        ApexCreateHelp: HELP_DEFAULTS + HELP_OUTPUTDIR_RELATIVE + HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'apex:class:create -n MyClass\n' +
            '   $ ' + CMD_HEAD + 'apex:class:create -n MyClass -d classes\n',
        ApexNameFlagDescription: NAME_FLAG_BASE + 'Apex class',
        ApexNameFlagLongDescription: 'The name of the new Apex class. ' + SF_NAME_RESTRICTIONS,
        ApexClassNamePrompt: 'What is the name of your Apex class?',
        ApexNameOption: 'The name of the Apex class',
        ApexTriggerCreateDescription: 'create an Apex trigger',
        ApexTriggerCreateLongDescription: 'Creates an Apex trigger' + LONG_DESCRIPTION_OUTPUTDIR +
            'If you don’t explicitly set the API version, it defaults to the current API version. ' +
            'The .trigger file and associated metadata file are created.',
        ApexTriggerCreateHelp: HELP_DEFAULTS + HELP_OUTPUTDIR_RELATIVE + HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'apex:trigger:create -n MyTrigger\n' +
            '   $ ' + CMD_HEAD + 'apex:trigger:create -n MyTrigger -s Account -e \'before insert, after upsert\'\n' +
            '   $ ' + CMD_HEAD + 'apex:trigger:create -n MyTrigger -d triggers\n',
        ApexTriggerNameFlagDescription: NAME_FLAG_BASE + 'Apex trigger',
        ApexTriggerNameFlagLongDescription: 'The name of the new Apex trigger. ' + SF_NAME_RESTRICTIONS,
        ApexTriggerSobjectFlagDescription: 'sObject to create a trigger on',
        ApexTriggerSobjectFlagLongDescription: 'The sObject to create an Apex trigger on.',
        ApexTriggerEventsFlagDescription: 'events that fire the trigger',
        ApexTriggerEventsFlagLongDescription: 'The events that cause the trigger to fire.',
        ApexTriggerNamePrompt: 'What is the name of your Apex trigger?',
        ApexTriggerNameOption: 'The name of the Apex trigger',
        ApexTriggerSobjectOption: 'The sObject for the Apex trigger',
        ApexTriggerTriggerEventsOption: 'The trigger events that activate the Apex trigger',
        InvalidTriggerEvent: '\'%s\' is not a valid trigger event.',
        LightningCreateDescription: 'create a Lightning %s',
        LightningCreateLongDescription: 'Creates a Lightning %s bundle' + LONG_DESCRIPTION_OUTPUTDIR +
            'The bundle consists of multiple files in a folder with the designated name.',
        LightningCreateHelp: HELP_DEFAULTS + HELP_OUTPUTDIR_RELATIVE + HELP_OUTPUTDIR_RELATIVE_LIGHTNING +
            HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s\n' +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s -d aura\n',
        LightningCmpCreateDescription: 'create a bundle for an Aura component or a Lightning web component',
        LightningCmpCreateLongDescription: 'Creates a bundle for an Aura component or a Lightning web component.'
            + LONG_DESCRIPTION_OUTPUTDIR +
            'The bundle consists of multiple files in a folder with the designated name.',
        LightningCmpCreateHelp: HELP_DEFAULTS + HELP_OUTPUTDIR_RELATIVE + HELP_OUTPUTDIR_RELATIVE_LIGHTNING +
            '\nTo create a Lightning web component, pass --type lwc to the command. ' +
            'If you don’t include a --type value, Salesforce CLI creates an Aura component by default.' +
            HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s\n' +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s --type lwc\n' +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s -d aura\n' +
            '   $ ' + CMD_HEAD + 'lightning:%s:create -n my%s --type lwc -d lwc\n',
        LightningNameFlagDescription: NAME_FLAG_BASE + 'Lightning %s',
        LightningNameFlagLongDescription: 'The Lightning %s name. ' + SF_NAME_RESTRICTIONS,
        BundleNameArgument: 'The name of the Lightning component',
        BundleTypeArgument: 'The type of the Lightning bundle',
        LightningAppChoice: 'New Lightning Application',
        LightningCmpChoice: 'New Lightning Component',
        LightningLwcChoice: 'New Lightning Web Component',
        LightningEvtChoice: 'New Lightning Event',
        LightningInfChoice: 'New Lightning Interface',
        LightningBundNamePrompt: 'What is the name of your Lightning bundle?',
        LightningBundTypePrompt: 'What type of Lightning bundle do you want to create?',
        InvalidBundleType: 'invalid bundle type for reflection',
        InvalidAuraBundlePath: 'Lightning bundles must have a parent folder named \'aura\'.',
        InvalidLwcBundlePath: 'Lightning bundles must have a parent folder named \'lwc\'.',
        LightningTypeFlagDescription: 'type of the Lightning component',
        LightningTypeFlagLongDescription: 'The type of the new Lightning component.',
        LightningInternalFlagDescription: 'create the Lightning component for internal use',
        LightningInternalFlagLongDescription: 'Creates the Lightning component for internal Salesforce use. ' +
            'Removes the restriction on parent folder name being lwc or aura and does not create -meta.xml files.',
        LightningTestCreateDescription: 'create a Lightning test',
        LightningTestCreateLongDescription: 'Creates a Lightning test' + LONG_DESCRIPTION_OUTPUTDIR +
            'The .resource file and associated metadata file are created.',
        LightningTestCreateHelp: HELP_OUTPUTDIR_RELATIVE + HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'lightning:test:create -n MyLightningTest\n' +
            '   $ ' + CMD_HEAD + 'lightning:test:create -n MyLightningTest -d lightningTests\n',
        LightningTestNameFlagDescription: NAME_FLAG_BASE + 'Lightning test',
        LightningTestNameFlagLongDescription: 'The name of the new Lightning test. ' + SF_NAME_RESTRICTIONS,
        LightningTestNamePrompt: 'What is the name of your Lightning test?',
        LightningTestNameOption: 'The name of the Lightning test',
        VfCreateDescription: 'create a Visualforce %s',
        VfCreateLongDescription: 'Creates a Visualforce %s' + LONG_DESCRIPTION_OUTPUTDIR +
            'The command creates the .%s file and associated metadata file.',
        VfCreateHelp: HELP_DEFAULTS + HELP_OUTPUTDIR_RELATIVE + 'Name and label are required.\n' +
            HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'visualforce:%s:create -n my%s -l mylabel\n' +
            '   $ ' + CMD_HEAD + 'visualforce:%s:create -n my%s -l mylabel -d %ss\n',
        VfNameFlagDescription: NAME_FLAG_BASE + 'Visualforce %s',
        VfNameFlagLongDescription: 'The Visualforce %s name. ' + SF_NAME_RESTRICTIONS,
        VfLabelFlagDescription: 'Visualforce %s label',
        VfLabelFlagLongDescription: 'The label saved in the metadata for the Visualforce %s.',
        VfNameOption: 'The name of the Visualforce ',
        VfLabelOption: 'The label string for the ',
        // Artifact* messages are not yet in use
        ArtifactNameFlagDescription: NAME_FLAG_BASE + 'artifact',
        ArtifactNameFlagLongDescription: 'TBD',
        ArtifactVersionFlagDescription: 'starting version number of the new artifact',
        ArtifactVersionFlagLongDescription: 'TBD',
        ArtifactDescriptionFlagDescription: 'description of the new artifact',
        ArtifactDescriptionFlagLongDescription: 'TBD',
        ArtifactMakeDefaultFlagDescription: 'TBD',
        ArtifactMakeDefaultFlagLongDescription: 'TBD',
        ArtifactCreateCommandDescription: 'create a new artifact',
        ArtifactCreateCommandLongDescription: 'TBD',
        ArtifactCreateCommandHelp: 'TBD',
        ArtifactNameOption: 'The name of the Artifact',
        ArtifactVersionOption: 'The version of the Artifact',
        ArtifactDescriptionOption: 'The descriptive text of the Artifact',
        ArtifactMakeDefaultOption: 'TBD',
        ArtifactNamePrompt: 'What is the name of your Artifact?',
        // ScratchDef* messages - none of these are exposed to the user
        ScratchDefNameFlagDescription: NAME_FLAG_BASE + 'scratch org definition',
        ScratchDefNameFlagLongDescription: 'TBD',
        ScratchDefCompanyFlagDescription: 'company of scratch org owner',
        ScratchDefCompanyFlagLongDescription: 'TBD',
        ScratchDefLastNameFlagDescription: 'last name of scratch org owner',
        ScratchDefLastNameFlagLongDescription: 'TBD',
        ScratchDefEmailFlagDescription: 'email of scratch org owner',
        ScratchDefEmailFlagLongDescription: 'TBD',
        ScratchDefCreateCommandDescription: 'create a new scratch org definition',
        ScratchDefCreateCommandLongDescription: 'TBD',
        ScratchDefCreateCommandHelp: 'TBD',
        ScratchDefNameOption: 'The scratch definition file name',
        ScratchDefCompanyOption: 'The company for the scratch def',
        ScratchDefLastNameOption: 'The last name of the scratch org owner',
        ScratchDefEmailOption: 'The email address of the scratch org owner',
        ScratchDefNamePrompt: 'What is the name of the new scratch definition file?',
        // Manifest - none of these are exposed to the user
        ManifestNameFlagDescription: NAME_FLAG_BASE + 'manifest',
        ManifestNameFlagLongDescription: '<>',
        ManifestCreateCommandDescription: 'create a new manifest',
        ManifestCreateCommandLongDescription: '<>',
        ManifestCreateCommandHelp: '<>',
        ManifestNameOption: 'The manifest file name',
        ManifestNamePrompt: 'What is the name of the new manifest file?',
        // Prettier - none of these are exposed to the user
        PrettierCreateCommandDescription: 'create prettier config',
        PrettierCreateCommandLongDescription: '<>',
        PrettierCreateCommandHelp: '<>',
        // VSCode - none of these are exposed to the user
        VSCodeCreateCommandDescription: 'create vscode config',
        VSCodeCreateCommandLongDescription: '<>',
        VSCodeCreateCommandHelp: '<>',
        VSCodePrettierFlagDescription: 'include recomended Prettier configuration',
        VSCodePrettierFlagLongDescription: 'Includes the recommended configuration for the Prettier code formatter.',
        VSCodeESLintFlagDescription: 'include recomended ESLint configuration',
        VSCodeESLintFlagLongDescription: 'Includes the recommended configuration for ESLint.',
        // ESLint - none of these are exposed to the user
        ESLintCreateCommandDescription: 'create eslint config',
        ESLintCreateCommandLongDescription: '<>',
        ESLintCreateCommandHelp: '<>',
        ProjectNameFlagDescription: NAME_FLAG_BASE + 'project',
        ProjectNameFlagLongDescription: 'The name for the new project. Any valid folder name is accepted.',
        ProjectTemplateFlag: 'template to use for project creation',
        ProjectTemplateFlagLongDescription: 'The template to use to create the project. ' +
            'Supplied parameter values or default values are filled into a copy of the template.',
        ProjectVersionFlagDescription: 'source API version number',
        ProjectVersionFlagLongDescription: 'The source API version for the project. ' +
            'The version defaults to the current release and will normally be defaulted.' +
            'This is used to properly import or export metadata.',
        ProjectLoginFlagDescription: 'Salesforce instance login URL',
        ProjectLoginFlagLongDescription: 'The login URL for the Salesforce instance being used.' +
            'Normally defaults to https://login.salesforce.com.',
        ProjectNamespaceFlagDescription: 'project associated namespace',
        ProjectNamespaceFlagLongDescription: 'The namespace associated with this project and any connected scratch orgs.',
        ProjectManifestFlagDescription: 'generate a manifest (package.xml) for change-set-based development',
        ProjectManifestFlagLongDescription: 'Generates a default manifest (package.xml) for fetching Apex, Visualforce, ' +
            'Lightning web components, Aura components, and static resources.',
        ProjectPackageFlagDescription: 'default package directory name',
        ProjectPackageFlagLongDescription: 'The default package directory name. ' +
            'Metadata items such as classes and Lightning bundles are placed inside this folder.',
        ProjectCreateCommandDescription: 'create a Salesforce DX project',
        ProjectCreateCommandLongDescription: 'Creates a Salesforce DX project ' + LONG_DESCRIPTION_OUTPUTDIR +
            'The command creates the necessary configuration files and folders.',
        ProjectCreateCommandHelp: 'Default values are used if the template, namespace, defaultpackagedir, and outputdir aren’t supplied.\n' +
            HELP_OUTPUTDIR_RELATIVE + HELP_EXAMPLES +
            '   $ ' + CMD_HEAD + 'project:create --projectname mywork\n' +
            '   $ ' + CMD_HEAD + 'project:create --projectname mywork --defaultpackagedir myapp\n' +
            '   $ ' + CMD_HEAD + 'project:create --projectname mywork --defaultpackagedir myapp --manifest\n' +
            '   $ ' + CMD_HEAD + 'project:create --projectname mywork --template empty',
        NamespaceLengthExceeded: 'Namespace can\'t be more than %s characters long.',
        ArtifactNameLengthExceeded: 'Name can\'t be more than %s characters long.',
        ProjectNameOption: 'The project root directory name',
        ProjectVersionOption: 'The version of the API to use',
        ProjectLoginOption: 'The URL of the Salesforce instance',
        ProjectDefaultArtifactOption: 'The default package directory to create',
        ProjectNamespaceOption: 'The project associated namespace',
        ProjectScratchOrgOption: 'The default scratch org definition',
        ProjectManifestOption: 'The default manifest definition',
        ProjectPrettierOption: 'Prettier recommended configuration',
        ProjectESLintOption: 'ESLint recommended configuration',
        ProjectNamePrompt: 'What is the name of your project?',
    }
};
